### PR TITLE
Call a variant in a selection article directly

### PR DESCRIPTION
**Pull request by heural**
It should be possible to preselect a variant in a selection article directly with a URL-Param e.g. "productId".
Maybe a use case is a product feed for various tag managers like google tag manager to expose all the variant products of a selection product.

### CODE REVIEW

File: templates/client/html/catalog/detail/body.php [hunk 1]
```diff
@@ -620,3 +620,30 @@
...
+            $js = '';
+            foreach($attributes as $attribId => $attribute):
+                $js .= '"' .  $attribute->getType() . '": "' . $attribute->getId() . '",
+                ';
+            endforeach;
+
+            if (!empty($js)): ?>
+               <script>
+                   AimeosVariantSelectionProduct = {"prodId": "<?= $this->detailProductItem->getId() ?>", "attributes": {<?= $js ?>}};
+               </script>
```
Comment (aimeos) [inline templates/client/html/catalog/detail/body.php:642]: Adding Javascript in the template won't work if Content-Security-Policy is enabled, which is done by default in most installations. You have to add that to a "data-..." attribute of an element instead.

File: templates/client/html/catalog/detail/body.php [hunk 2]
```diff
@@ -620,3 +620,30 @@
 	</div>
 
 <?php endif ?>
+
+<?php
+// preselection of a variant in a selection article
+if (isset($_GET['productId']) && is_numeric($_GET['productId'])):
```
Comment (aimeos) [inline templates/client/html/catalog/detail/body.php:626]: Don't use `$_GET`, use `$this->param( 'productid' )` instead

File: templates/client/html/catalog/detail/body.php [hunk 3]
```diff
@@ -620,3 +620,30 @@
...
+
+    foreach( $this->detailProductItem->getRefItems( 'product', null, 'default' ) as $prodid => $product ) :
+
+        if ($prodid == $_GET['productId']):
```
Comment (aimeos) [inline templates/client/html/catalog/detail/body.php:630]: You can get the product much simpler by using:
```
if( ( $listItem = $this->detailProductItem->getListItem( 'product', 'default', $prodId ) ) && ( $product = $listItem->getRefItem() ) )
```

File: templates/client/html/catalog/detail/body.php [hunk 4]
```diff
@@ -620,3 +620,30 @@
...
+
+            $attributes = $product->getRefItems('attribute',null,'variant');
+            $js = '';
+            foreach($attributes as $attribId => $attribute):
+                $js .= '"' .  $attribute->getType() . '": "' . $attribute->getId() . '",
+                ';
+            endforeach;
```
Comment (aimeos) [inline templates/client/html/catalog/detail/body.php:637]: Don't build a JSON-like string yourself but create an array of key/value pairs and use `json_encode()` afterwards

**Comment by aimeos** [comment]
Thank you for your PR!
There are a few problems in your code which must be fixed first before we can merge your PR (see comments)

File: templates/client/html/catalog/detail/body.php [hunk 5]
```diff
@@ -48,7 +48,27 @@
 
 
 ?>
-<?php if( isset( $this->detailProductItem ) ) : ?>
+<?php if( isset( $this->detailProductItem ) ) : 
+
+        // preselection of a variant product in a selection article
+        $preSelectVariantAttributes = [];
+        if ($_GET['productId'] && is_numeric($_GET['productId'])):
```
Comment (aimeos) [inline templates/client/html/catalog/detail/body.php:55]: Don't use `$_GET`, but `$this->param( 'productId' )` instead

File: templates/client/html/catalog/detail/body.php [hunk 6]
```diff
@@ -48,7 +48,27 @@
...
+
+                    $attributes = $product->getRefItems('attribute',null,'variant');
+
+                    foreach($attributes as $attribId => $attribute):
+                        $preSelectVariantAttributes[$attribute->getType()] =  $attribute->getId();
+                    endforeach;
+
+                    if (!empty($preSelectVariantAttributes)):
+                        $preSelectVariantAttributes = htmlspecialchars(json_encode(['attributes' => $preSelectVariantAttributes, 'prodId' => $this->detailProductItem->getId()]), ENT_QUOTES, 'UTF-8');
+                    endif;
```
Comment (aimeos) [inline templates/client/html/catalog/detail/body.php:67]: You don't need this, `$enc->attr()` will do the correct encoding and escaping for you

File: templates/client/html/catalog/detail/body.php [hunk 7]
```diff
@@ -58,7 +78,8 @@
 			<!-- catalog.detail.navigator -->
 
 			<article class="product row <?= $this->detailProductItem->getConfigValue( 'css-class' ) ?>"
-				data-id="<?= $this->detailProductItem->getId() ?>" data-reqstock="<?= $reqstock ?>">
+				data-id="<?= $this->detailProductItem->getId() ?>" data-reqstock="<?= $reqstock ?>"
+				<?= !empty($preSelectVariantAttributes) ? 'data-preselectvariant="' . $preSelectVariantAttributes . '"' : '' ?>>
```
Comment (aimeos) [inline templates/client/html/catalog/detail/body.php:82]: Just write:
```
data-preselectvariant="<?= $enc->attr( $preSelectVariantAttributes ) ?>"
```

File: themes/client/html/default/aimeos.js [hunk 8]
```diff
@@ -446,15 +446,11 @@ AimeosBasket = {
...
-				$.each(this.attributes, function(key,val){
-					$('#select-' + prodId + '-' + key).val(val).trigger('change');
-				});
-			});
-
+		if ($('article.product.row').data('preselectvariant')) {
+		       let preSelected = $('article.product.row').data('preselectvariant');
+		       $.each(preSelected.attributes, function (key, val) {
+			   $('#select-' + preSelected.prodId + '-' + key).val(val).trigger('change');
+		       });
```
Comment (aimeos) [inline themes/client/html/default/aimeos.js:453]: Better would be:
```
const product = $('article.product');
if (product && product.data('preselectvariant')) {
  // ...
}
```
Also, do you we really need the `attributes` sub-key? We can get the product ID from the same node using `product.data('id')`

File: templates/client/html/catalog/detail/body.php (hunk 5 above)
Comment (heural) [reply templates/client/html/catalog/detail/body.php:55]: My url looks like this:
`http://aimeos.local/detail/d/Testproduct-Selection?productId=19963`
and there is no $this->param( 'productId' ) in body.php.
I think I have to configure a url-param like
`http://aimeos.local/detail/d/Testproduct-Selection/0/19963`

Please can you assist me where I have to do that?

Comment (heural) [reply templates/client/html/catalog/detail/body.php:55]: Ok, I have foud it ;-)
https://aimeos.org/docs/2022.x/typo3/setup/#seo-urls

File: templates/client/html/catalog/detail/body.php [hunk 9]
```diff
@@ -48,7 +48,23 @@
 
 
 ?>
-<?php if( isset( $this->detailProductItem ) ) : ?>
+<?php if( isset( $this->detailProductItem ) ) : 
+
+        // preselection of a variant product in a selection article
+        $preSelectVariantAttributes = [];
+        if ($this->param( 'productId' ) && is_numeric($this->param( 'productId' ))):
```
Comment (aimeos) [inline templates/client/html/catalog/detail/body.php:55]: The product ID doesn't has to be numeric, that's only currently the case in relational databases

File: templates/client/html/catalog/detail/body.php [hunk 10]
```diff
@@ -48,7 +48,23 @@
...
+
+            if(($listItem = $this->detailProductItem->getListItem( 'product', 'default', (int) $this->param( 'productId' ))) && ( $product = $listItem->getRefItem() ) )
```
Comment (aimeos) [inline templates/client/html/catalog/detail/body.php:57]: `getListItem()` requires a string as last parameter

File: templates/client/html/catalog/detail/body.php [hunk 11]
```diff
@@ -48,7 +48,23 @@
...
+
+                    $attributes = $product->getRefItems('attribute',null,'variant');
+
+                    foreach($attributes as $attribId => $attribute):
+                        $preSelectVariantAttributes[$attribute->getType()] =  $attribute->getId();
+                    endforeach;
```
Comment (aimeos) [inline templates/client/html/catalog/detail/body.php:63]: You can get this easier by using:
```
$varAttr = $product->getRefItems( 'attribute', null, 'variant' )->col( 'attribute.id', 'attribute.type' );
```

**Comment by aimeos** [comment]
Looks good now, thank you very much! :-)

**Pull request merged**